plugins/module_utils/foreman_helper.py

```diff
@@ -364,6 +462,34 @@ def sanitize_entity_dict(entity_dict, name_map):
     return {value: entity_dict[key] for key, value in name_map.items() if key in entity_dict}
 
 
+def _flatten_entity(entity, entity_spec):
+    """Flatten entity according to spec"""
+    result = {}
+    for key, value in entity.items():
+        if key in entity_spec:
```

Comment:
Won't this remove all keys that are already in the flattened name? (When you pass an existing entity)

Reply:
you mean, that you cannot 'flatten' it twice? That would be correct. But that nothing new compared to the existing sanitize name_map mechanism.

Reply:
But did we ever run sanitize on the data received from the api? We only did flatten_value before, and that did not take into account the map.

Reply:
We could make it idempotent by adding all the flat_names as keys with empty dictionary. Do you think it is worth?

Reply:
Well, I think it's mandatory. But maybe I'm also missreading the code -- only on my mobile at the moment.

You call flatten_entity on current_entity in update, and current contains the api generated answer. So after flatten it will be missing entries. And then you compare it to desired_entity and stuff is missing. Right?

Reply:
Not as such. We only call flatten once on either entities. But yes, if the api answer contained e.g. `fullname` and we decide to call it `description` (for whatever reason?!?) we'd end up with something like 
https://github.com/theforeman/foreman-ansible-modules/blob/f169172326368851d4836f0d413b9898fcdb2d25/plugins/modules/foreman_domain.py#L110-L111

For referenced entities, we get the hydrated nested (thin) entities in the api answer, like `domain: {id:1}`, therefore we can employ the same flattening mechanism there.

Reply:
OK

Reply:
And in cas we do not get the hydrated version from the api, it looks like:
https://github.com/theforeman/foreman-ansible-modules/blob/c88ffc2e10a8e7a486421ce22f62407bb24c0009/plugins/modules/foreman_hostgroup.py#L138-L139

Reply:
I think we can make that nicer, but that would make this PR even bigger, so let's discuss that in #357